components/calendar/src/error.rs

```diff
@@ -59,13 +63,109 @@ pub enum DateError {
+    Range(RangeError),
+    /// The era code is invalid for the calendar.
+    #[displaydoc("Unknown era or invalid syntax")]
+    UnknownEra,
+    /// The month code syntax is invalid.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::error::DateFromFieldsError;
+    /// use icu_calendar::Iso;
+    /// use icu_calendar::types::MonthCode;
+    /// use icu_calendar::types::DateFields;
+    /// use tinystr::tinystr;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.extended_year = Some(2000);
+    /// fields.month_code = Some(MonthCode(tinystr!(4, "????")));
+    /// fields.day = Some(1);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields, Default::default(), Iso
+    /// )
+    /// .expect_err("month code is invalid");
+    ///
+    /// assert!(matches!(err, DateFromFieldsError::InvalidMonthCode));
+    /// ```
+    #[displaydoc("Invalid month code syntax")]
+    InvalidMonthCode,
+    /// The specified month code does not exist in this calendar.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::error::DateFromFieldsError;
+    /// use icu_calendar::cal::Hebrew;
+    /// use icu_calendar::types::DateFields;
+    /// use icu_calendar::types::MonthCode;
+    /// use tinystr::tinystr;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.extended_year = Some(5783);
+    /// fields.month_code = Some(MonthCode::new_normal(13).unwrap());
+    /// fields.day = Some(1);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields,
+    ///     Default::default(),
+    ///     Hebrew
+    /// )
+    /// .expect_err("no month M13 in Hebrew");
+    /// assert!(matches!(err, DateFromFieldsError::UnknownMonthCodeForCalendar(_)));
+    /// ```
+    #[displaydoc("Month code does not exist in this calendar: {0:?}")]
+    UnknownMonthCodeForCalendar(MonthCode),
+    /// The specified month code exists in this calendar, but not in the specified year.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::error::DateFromFieldsError;
+    /// use icu_calendar::cal::Hebrew;
+    /// use icu_calendar::types::DateFields;
+    /// use icu_calendar::types::MonthCode;
+    /// use tinystr::tinystr;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.extended_year = Some(5783);
+    /// fields.month_code = Some(MonthCode::new_leap(5).unwrap());
+    /// fields.day = Some(1);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields,
+    ///     Default::default(),
+    ///     Hebrew
+    /// )
+    /// .expect_err("no month M05L in Hebrew year 5783");
+    /// assert!(matches!(err, DateFromFieldsError::UnknownMonthCodeForYear(_)));
+    /// ```
+    #[displaydoc("Month code exists in calendar, but not for this year: {0:?}")]
+    UnknownMonthCodeForYear(MonthCode),
```

Comment:
it's not unknown, it's invalid

Reply:
CC @Manishearth for feedback on naming of the error variant